### PR TITLE
Anonymous proposals for registered users

### PR DESCRIPTION
**Pull request by entantoencuanto**
Fixes PopulateTools/issues#1894

This PR adds the following changes to the anonymous proposals feature:

* [x] Allows signed in users to create, update and publish anonymous proposals. Once the proposal is created as anonymous the draft edition is allowed but the authorship cannot be changed. Also an existing proposal with an author can't be transformed to anonymous (PopulateTools/decidim-module-anonymous_proposals#14)
* [x] Changes the default value of anonymous proposals setting and disabling the feature on new proposals components (PopulateTools/decidim-module-anonymous_proposals#13)
* [x] Adds some javascript to keep the title and body of new proposals recovering the value when the user is signed in registered, or signed out. The change affects to plain text body input and rich text editor if enabled (PopulateTools/decidim-module-anonymous_proposals#15)

Test the feature in https://decidim-dev-27.populate.tools/processes/new-process/f/14/ and select a municipality scope for the proposal, some bug is preventing the creation of proposals in the proposals components of the other participatory process (something related with the absence of a scope in the space)

[Screencast from 31-01-24 19:05:29.webm](https://github.com/PopulateTools/decidim-module-anonymous_proposals/assets/446459/a52c1b11-00d1-4215-b809-85d3999ab42c)

### CODE REVIEW

**Comment by entantoencuanto** [comment]
Oh, I'm sorry, I added a bug in the process of integration with simple proposals and Decidim awesome and actually the error comes from the other [PR](https://github.com/PopulateTools/decidim-module-anonymous_proposals/pull/18) in process. I've fixed the bug there, but I'm going to deploy again the application with the changes introduced by this PR (those changes don't include the simple proposals feature activation, so you will see the standard proposal creation flow with multiple steps)

Sorry for the inconveniences, can you check again, please?

**Comment by simonaszilinskas** [comment]
image

Hello, thank you for the fast reaction. Now I don't have access to the anonymous group anymore :( let me know when I can test it out again. Thanks!

**Comment by entantoencuanto** [comment]
With which user are you trying to create anonymous proposals? Maybe the issue is with the component configuration. With admin@example.org I have access to the "Anonymous user" option https://decidim-dev-27.populate.tools/processes/wear-exaggerate/f/3/proposals/new (there is no select if the user is not signed in because then the proposal can be only created anonymously)

There may be other problem related with the standard behavior of the application. If a user has an existing draft proposal the new proposal redirects him to the draft edition instead of the new form.

I have added also a rule to prevent changing a proposal from anonymous to having an author and vice versa once the proposal is created, so currently the only step where a signed in user can select between anonymous and the rest of identities for proposal authorship is in the new step.

**Comment by simonaszilinskas** [comment]
Thanks for the response @entantoencuanto !

Here is the screenshot of the current situation: the user is signed in and the anonymous group option isn't there. 

Intended situation:
As a signed-in user, when creating a proposal in a component with anonymous proposals activated, I want to have the option "Anonymous" in the "Create proposal as..." selector, so that I can submit anonymous proposals even when signed-in.

**Comment by entantoencuanto** [comment]
In the screenshot you show me the proposal has been created but not yet published. I added the condition to not allow the user to change the authorship of the proposal between anonymous or not once created, because with the changes of this PR an anonymous proposal can be edited by any signed in user (while the proposal is editable), so an anonymous proposal could be appropriated by a user knowing its edit URL in this phase.

I can make changes to allow users to change authorship at any moment while the proposal can be edited but preventing the change from anonymous to authored and vice versa if the proposal has already been published.

Or do you prefer to allow changes in authorship type at any moment if the proposal is editable?

**Comment by simonaszilinskas** [comment]
I think it is very good to allow users to change authorship at any moment while the proposal can be edited but preventing the change from anonymous to authored and vice versa if the proposal has already been published knowing that if we allow the edit post publishing someone else could change authorship (that we don't want).

**Comment by simonaszilinskas** [comment]
Perfect, thank you @entantoencuanto ! Is the latest version available on your staging platform?

**Comment by entantoencuanto** [comment]
Yes, you can check the changes here: https://decidim-dev-27.populate.tools/processes/wear-exaggerate/f/3/proposals/new

For the moment the module is installed alone without simple proposals and decidim awesome, if the review is OK we can check it with the other installed

**Comment by entantoencuanto** [comment]
Hi @simonaszilinskas I have installed in the staging application the simple proposal and the adapted awesome (from https://github.com/octree-gva/decidim-module-decidim_awesome) modules to check the compatibility

**Comment by simonaszilinskas** [comment]
https://github.com/PopulateTools/decidim-module-anonymous_proposals/assets/14858948/2ffa153a-9c40-490e-9329-195f9ca7a503

Hello @entantoencuanto thank you very much, it looks great! just two considerations present in the video: 
- 00:43 - the banner to log in or sign up to assign to an account is not present. If that's not too hard of an add, it would be good to have it there;
- 00:54 - server issue

The rest seems good from my perspective. I will write the test cases and then send it to the rest of our team to review, but it looks good

**Comment by entantoencuanto** [comment]
Hi, I have added a fix to avoid the server error (I thought it was a Decidim installation error but it was actually a bug in the module) and changes to include the announcement banner in the edit_draft/complete page for not signed in users

The changes are deployed in the staging application: https://decidim-dev-27.populate.tools/processes/wear-exaggerate/f/3/proposals/new

**Comment by furilo** [comment]
@simonaszilinskas are you OK with this? Anything missing?

**Comment by furilo** [comment]
Hi @simonaszilinskas - Edu is off a few days because of personal matters. Give us some time until he is back. Thanks!

**Review by moustachu — APPROVED** [review]

**Comment by entantoencuanto** [comment]
Hi, @simonaszilinskas, do you mean to disable the edition of anonymous proposals after its publication or even also after its creation? I can include a restriction

About the letter_opener configuration I'm researching the way

**Comment by simonaszilinskas** [comment]
> Hi, @simonaszilinskas, do you mean to disable the edition of anonymous proposals after its publication or even also after its creation? I can include a restriction
> 
> About the letter_opener configuration I'm researching the way

Hi @entantoencuanto , could you please elaborate a bit more on the first question - I'm not sure I understood it. I think the crucial need is not to have cases where other people other than the creator of a proposal can edit a proposal. This is currently the case for anonymous proposals.

**Comment by entantoencuanto** [comment]
The creation of an anonymous proposal has 3 phases:

* The initial step when the user is in the page with the form to create a new proposal. At this stage the proposal does not exist yet and only the user can edit the proposal and create it by submitting the form.
* After submitting the form the proposal is created as draft. The draft proposal is not shown in the lists of proposals but can be edited if the user knows the URL.
* Finally the proposal can be published and there is a setting in the proposal component which allows the user to edit the proposals after its creation. This configuration can be disabled so once the proposal is published nobody can edit a proposal.

So the application can prevent the edition of anonymous proposals once published without any additional change introduced by the module. Or we should prevent the edition of anonymous proposals regardless the component settings and the draft or published state of the proposals? (to get this some additional changes has to be included by the module)

**Comment by simonaszilinskas** [comment]
Thanks for the super clear comment. I think the draft state is a very minor risk, but once published, anonymous proposals should not be editable if someone else can see them in the list and can edit them. Ideally, the "normal" proposals should still follow the editability setting you mention.

**Comment by entantoencuanto** [comment]
Hi, I have added a change to prevent the edition of proposals once published. The changes can be checked at https://decidim-dev-27.populate.tools/processes/wear-exaggerate/f/3/proposals/

**Comment by simonaszilinskas** [comment]
Thank you @entantoencuanto ! looks great !

**Comment by entantoencuanto** [comment]
Hi, @simonaszilinskas, I have prepared a small branch to install this module in your [app](https://github.com/OpenSourcePolitics/decidim-app). Please, could you provide me access to the repository to create a PR?. The branch used compatible with the 0.27 Decidim release is the one used in this PR: `anonymous_proposals_for_registered_users`.

The PR basically adds the dependency into the Gemfile. The next steps to make work the module are:
1. Run the `decidim_anonymous_proposals:generate_anonymous_group` task to create an anonymous group to associate the proposals with
2. Set the DEFACE_ENABLED environment variable as `true` to allow the module to ovewrite the forms and create new proposal button
3. Enable the feature in a proposals component of the application

**Comment by simonaszilinskas** [comment]
@entantoencuanto could you just share the link to the fork you made for the PR ? I'm waiting for the approval of our team to grand access for the PR

**Comment by entantoencuanto** [comment]
PR created https://github.com/OpenSourcePolitics/decidim-app/pull/517

**Pull request merged**
